daemon/albatross_console.ml

```diff
@@ -116,41 +128,86 @@ let open_fifo name =
         Logs.err (fun m -> m "%a error while reading %s" Fpath.pp fifo (Printexc.to_string exn)) ;
         Lwt.return None)
 
-let t = ref Vmm_core.String_map.empty
-
 let fifos = Vmm_core.conn_metrics "fifo"
 
-let add_fifo id =
+let add_fifo n id =
   let name = Vmm_core.Name.to_string id in
-  open_fifo id >|= function
-  | None -> Error (`Msg "opening")
+  let path = Vmm_core.Name.drop_label id in
+  let lbl = Option.get (Vmm_core.Name.name id) in
+  open_fifo id >>= function
+  | None -> Lwt.return (Error (`Msg "opening"))
   | Some f ->
-    let ring = match Vmm_core.String_map.find_opt name !t with
-      | None ->
-        let ring = Vmm_ring.create "" () in
-        let map = Vmm_core.String_map.add name ring !t in
-        t := map ;
-        ring
-      | Some ring -> ring
-    in
+    (match Vmm_trie.find path !state with
+     | None ->
+       let ringbuffer = Vmm_ring.create "" () in
+       let map = LMap.singleton lbl ([], true, ringbuffer) in
+       let trie, _ = Vmm_trie.insert path (n, map) !state in
+       state := trie;
+       Lwt.return ringbuffer
+     | Some (_, map) ->
+       match LMap.find_opt lbl map with
+       | None ->
+         let ringbuffer = Vmm_ring.create "" () in
+         let map = LMap.add lbl ([], true, ringbuffer) map in
+         (if LMap.cardinal map > n then
+            match LMap.choose_opt (LMap.filter (fun _ (_, active, _) -> not active) map) with
+            | Some (key, (fds, _, _)) ->
+              Logs.warn (fun m -> m "dropping %s:%s"
+                            (Vmm_core.Name.to_string path)
+                            (Vmm_core.Name.Label.to_string key));
+              Lwt_list.iter_p Vmm_lwt.safe_close (List.map snd fds) >|= fun () ->
+              LMap.remove key map
+            | None ->
+              Logs.err (fun m -> m "intended to drop something while adding ringbuffer %s, but found nothing inactive" name);
+              Lwt.return map
```

Comment:
I guess this can happen if we have a policy of N unikernels and later substitute the policy with N' < N unikernels. Or perhaps the "root" policy can override policies and add more unikernels? Can this become spammy in that case?